library/alloc/src/vec/spec_from_iter_nested.rs

```diff
@@ -22,21 +22,45 @@ where
         // empty, but the loop in extend_desugared() is not going to see the
         // vector being full in the few subsequent loop iterations.
         // So we get better branch prediction.
-        let mut vector = match iterator.next() {
-            None => return Vec::new(),
-            Some(element) => {
-                let (lower, _) = iterator.size_hint();
-                let initial_capacity =
-                    cmp::max(RawVec::<T>::MIN_NON_ZERO_CAP, lower.saturating_add(1));
-                let mut vector = Vec::with_capacity(initial_capacity);
-                unsafe {
-                    // SAFETY: We requested capacity at least 1
-                    ptr::write(vector.as_mut_ptr(), element);
-                    vector.set_len(1);
-                }
-                vector
-            }
+        let (low, high) = iterator.size_hint();
+        assert!(
+            high.is_none_or(|high| low <= high),
+            "size_hint ({low:?}, {high:?}) is malformed from iterator {} collecting into {}",
+            core::any::type_name::<I>(), core::any::type_name::<Self>(),
+        );
+
+        let Some(first) = iterator.next() else {
+            return Vec::new();
+        };
```

Comment:
`next` and then `size_hint` is better since some adapters can provide a better hint after the first step.